nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/SubtypesJsonEnricher.java

```diff
@@ -0,0 +1,112 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Charles Boidot <cboidot@nuxeo.com>
+ */
+
+package org.nuxeo.ecm.platform.types;
+
+import static org.nuxeo.ecm.core.io.registry.reflect.Instantiations.SINGLETON;
+import static org.nuxeo.ecm.core.io.registry.reflect.Priorities.REFERENCE;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
+import org.apache.commons.lang3.BooleanUtils;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.io.marshallers.json.enrichers.AbstractJsonEnricher;
+import org.nuxeo.ecm.core.io.registry.reflect.Setup;
+import org.nuxeo.ecm.core.schema.SchemaManager;
+import org.nuxeo.ecm.platform.types.localconfiguration.UITypesConfigurationConstants;
+import org.nuxeo.runtime.api.Framework;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+
+/**
+ * Enrich {@link DocumentModel} JSON object with an array of the document types that can be created under the current
+ * document taking account the local configuration.
+ *
+ * @since 11.5
+ */
+@Setup(mode = SINGLETON, priority = REFERENCE)
+public class SubtypesJsonEnricher extends AbstractJsonEnricher<DocumentModel> {
+
+    public static final String NAME = "subtypes";
+
+    public SubtypesJsonEnricher() {
+        super(NAME);
+    }
+
+    @Override
+    public void write(JsonGenerator jg, DocumentModel enriched) throws IOException {
+        SchemaManager schemaManager = Framework.getService(SchemaManager.class);
+        Collection<String> subtypes = computeSubtypes(enriched);
+        jg.writeFieldName(NAME);
+        jg.writeStartArray();
+        for (String subtype : subtypes) {
+            jg.writeStartObject();
+            jg.writeStringField("type", subtype);
+            jg.writeArrayFieldStart("facets");
+            for (String facet : schemaManager.getDocumentType(subtype).getFacets()) {
+                jg.writeString(facet);
+            }
+            jg.writeEndArray();
+            jg.writeEndObject();
+        }
+        jg.writeEndArray();
+    }
+
+    protected Collection<String> computeSubtypes(DocumentModel enriched) {
+        Collection<String> defaultSubtypes = enriched.getDocumentType().getAllowedSubtypes();
+        if (enriched.hasFacet(UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_FACET)) {
+            return computeLocalConfigurationSubtypes(enriched, defaultSubtypes);
+        }
+        return defaultSubtypes;
+    }
+
+    protected Collection<String> computeLocalConfigurationSubtypes(DocumentModel enriched,
+            Collection<String> defaultSubtypes) {
+        Collection<String> subtypes = new HashSet<String>();
```

Comment:
You can use `Set` directly instead of Collection, furthermore, there's no need to declare the type in the HashSet implementation, this line would be:
```suggestion
        Set<String> subtypes = new HashSet<>();
```